pkg/controller/sparkapplication/sparkapp_metrics.go

```diff
@@ -0,0 +1,229 @@
+	validSpecLabels := make(map[string]string)
+	for labelKey, v := range specLabels {
+		newKey := util.CreateValidMetricNameLabel("", labelKey)
+		validSpecLabels[newKey] = v
+	}
+
+	metricLabels := make(map[string]string)
+	for _, label := range labels {
+		metricLabels[label] = validSpecLabels[label]
+		if metricLabels[label] == "" {
+			metricLabels[label] = "Unknown"
+		}
+	}
+	return metricLabels
+}
+
+func (sm *sparkAppMetrics) exportMetrics(oldApp, newApp *v1alpha1.SparkApplication) {
+	metricLabels := fetchMetricLabels(newApp.Labels, sm.Labels)
+	glog.V(2).Infof("Posting Metrics for %s. OldStatus: %v NewStatus: %v", newApp.GetName(), oldApp.Status, newApp.Status)
+
+	oldState := oldApp.Status.AppState.State
+	newState := newApp.Status.AppState.State
+	switch newState {
+	case v1alpha1.SubmittedState:
+		if isValidDriverStateTransition(oldState, newState) {
+			if m, err := sm.SparkAppSubmitCount.GetMetricWith(metricLabels); err != nil {
+				glog.Errorf("Error while posting metrics: %v", err)
+			} else {
+				m.Inc()
+			}
+		}
+	case v1alpha1.RunningState:
+		if isValidDriverStateTransition(oldState, newState) {
+			sm.SparkAppRunningCount.Inc(metricLabels)
+		}
+	case v1alpha1.CompletedState:
+		if isValidDriverStateTransition(oldState, newState) {
+			if !newApp.Status.SubmissionTime.Time.IsZero() && !newApp.Status.CompletionTime.Time.IsZero() {
+				d := newApp.Status.CompletionTime.Time.Sub(newApp.Status.SubmissionTime.Time)
+
+				if m, err := sm.SparkAppSuccessExecutionTime.GetMetricWith(metricLabels); err != nil {
+					glog.Errorf("Error while posting metrics: %v", err)
+				} else {
+					m.Observe(float64(d / time.Microsecond))
+				}
+			}
+			sm.SparkAppRunningCount.Dec(metricLabels)
+			if m, err := sm.SparkAppSuccessCount.GetMetricWith(metricLabels); err != nil {
+				glog.Errorf("Error while posting metrics: %v", err)
+			} else {
+				m.Inc()
+			}
+		}
+	case v1alpha1.FailedSubmissionState:
+		fallthrough
+	case v1alpha1.FailedState:
+		if isValidDriverStateTransition(oldState, newState) {
+			if !newApp.Status.SubmissionTime.Time.IsZero() && !newApp.Status.CompletionTime.Time.IsZero() {
+				d := newApp.Status.CompletionTime.Time.Sub(newApp.Status.SubmissionTime.Time)
+				if m, err := sm.SparkAppFailureExecutionTime.GetMetricWith(metricLabels); err != nil {
+					glog.Errorf("Error while posting metrics: %v", err)
+				} else {
+					m.Observe(float64(d / time.Microsecond))
+				}
+			}
+			sm.SparkAppRunningCount.Dec(metricLabels)
+			if m, err := sm.SparkAppFailureCount.GetMetricWith(metricLabels); err != nil {
+				glog.Errorf("Error while posting metrics: %v", err)
+			} else {
+				m.Inc()
+			}
+		}
+	}
+
+	// Potential Executor status updates
+	for executor, newExecState := range newApp.Status.ExecutorState {
+		if newExecState == v1alpha1.ExecutorRunningState && (isValidExecutorStateTransition(oldApp.Status.ExecutorState[executor], newExecState)) {
```

Comment:
Can we use a `switch` block that checks `newExecState` instead of embedding the check into the if blocks?

Reply:
done